### DISCUSSION TITLE
Including other Julia packages in "full_objective" definition

### DISCUSSION THREAD

**Discussion by mchiquier** [MilesCranmer/PySR#477] [General]
I am having trouble finding the documentation for this. If I wanted to use another external Julia package in the definition of my full_objective function, how would I do this (if it is possible)? I get errors if I place it before the function definition as well as within the function definition. Could you please point me to the right resources for this? Thanks!

**Comment by MilesCranmer** [comment]
Does this work? https://astroautomata.com/PySR/examples/#7-julia-packages-and-types

**Reply by mchiquier** [reply]
So I had actually tried that but it didn't work, though I'm sure I'm doing something wrong. I copied everything exactly as in the link above, and when I evaluate jl.p(1) outside of the model definition it works fine, but when I define it within the full_objective function definition, I get the following error: 

RuntimeError: <PyCall.jlwrap (in a Julia function called from Python)
JULIA: UndefVarError: `jl` not defined

and I should mention I also tried defining jl inside the string that contains the function definition but outside of the function just in case, and that didn't work either:

julia.core.JuliaError: Exception 'syntax: { } vector syntax is discontinued' occurred while calling julia code:

full_objective="""
                jl = pysr.julia_helpers.init_julia(julia_kwargs={"threads": 8, "optimize": 3})
                jl.eval("import Pkg;Pkg.add("Primes")")
                jl.eval("import Primes")
                function my_objective(tree::Node{T}, dataset::Dataset{T}, options::Options) where T 
                  prediction, flag = eval_tree_array(tree, dataset.X, options) 
                  !flag && return T(Inf) 
                  mystring = string_tree(tree, options)
                  count_sin = count(occursin("sin", mystring) for _ in eachindex(mystring))
                  #println("Number of occurrences of 'sin': ", count_sin)
                  jl.p(1)
                  abs3(x) = abs(x) 
                  #print(("loss: ", sum(abs3, prediction .- dataset.y) / length(prediction)))
                  return (sum(abs3, prediction .- dataset.y)) + count_sin 
              end"""

**Reply by MilesCranmer** [reply]
Oh, those lines should be run outside of that string. Those lines are regular Python code. So your entire script would be like

```python
import pysr
jl = pysr.julia_helpers.init_julia(julia_kwargs={"threads": 8, "optimize": 3})
jl.eval("import Pkg;Pkg.add("Primes")")
jl.eval("import Primes")
from pysr import PySRRegressor


model = PySRRegressor(...,
    full_objective="""
    function my_objective(tree::Node{T}, dataset::Dataset{T}, options::Options) where T
        prediction, flag = eval_tree_array(tree, dataset.X, options)
        !flag && return T(Inf)
        mystring = string_tree(tree, options)
        count_sin = count(occursin("sin", mystring) for _ in eachindex(mystring))
        #println("Number of occurrences of 'sin': ", count_sin)
        abs3(x) = abs(x)
        #print(("loss: ", sum(abs3, prediction .- dataset.y) / length(prediction)))
        return (sum(abs3, prediction .- dataset.y)) + count_sin
    end"""
)
```

Basically each call of `jl.eval` is a Python call that runs the Julia code (in the string). The `full_objective` is another string of Julia code – so you can't put Python code in it.


Also I don't think the `count(occursin("sin", mystring) for _ in eachindex(mystring))` is quite right. `eachindex` will give the indices of each character in the string. Probably easier to just do the method I gave on https://github.com/MilesCranmer/PySR/discussions/475 – which will also be much faster.

**Reply by mchiquier** [reply]
Well I need to call a function from the external package inside the my_objective function, and I can't quite work out how its done. In the example you've provided above, you aren't calling any function from the Primes package inside 'my_objective'.

And nice catch, will fix, thanks!

**Reply by mchiquier** [reply]
Ah, I just did: 

`import pysr
jl = pysr.julia_helpers.init_julia(julia_kwargs={"threads": 8, "optimize": 3})
jl.eval("import Pkg;Pkg.add("Primes")")
jl.eval("import Primes")
from pysr import PySRRegressor


model = PySRRegressor(...,
    full_objective="""
    function my_objective(tree::Node{T}, dataset::Dataset{T}, options::Options) where T
        prediction, flag = eval_tree_array(tree, dataset.X, options)
        !flag && return T(Inf)
        mystring = string_tree(tree, options)
        count_sin = count(occursin("sin", mystring) for _ in eachindex(mystring))
        #println("Number of occurrences of 'sin': ", count_sin)
        print(Primes.prime(round(Int, 7)), "test")
        #print(("loss: ", sum(abs, prediction .- dataset.y) / length(prediction)))
        return (sum(abs, prediction .- dataset.y)) + count_sin
    end"""
)`

and it worked, thanks!
